### ISSUE TITLE
Deferred implementation in C

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2245]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2245|
|Type|enhancement|
|Created|2006-11-21 16:37:26Z|
|Branch|https://github.com/twisted/twisted/tree/cdefer-2245-4|

For optimization (and fun) purpose I created a implementation of Deferred in C, inspired by the pyrex version lying around svn sandbox. I intentionally left out the debug and timeout stuff.

It seems to pass most Twisted suite, except those involving setTimeout and debug.

I'm putting this here for anyone interested. On some basic tests, my apps are 5 to 20% faster.

Attachments:

* [cdefer.c](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/82e61eb524a45c85864c61f2fb28ef76838dc4cc.c) (24072 bytes) - added by therve on 2006-11-21 16:38:05Z - 
* [deferbenchmark.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/65dc52a741222efb94f8d082cb3cf79f835e4b7a.patch) (2181 bytes) - added by Peaker on 2007-06-28 16:50:59Z - 
* [cdefer.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/bcd8af2ff0a34a69f4c2d5eb2c9bdc4264002246.patch) (33343 bytes) - added by Peaker on 2007-06-28 21:24:23Z - After my review -- fixing a lot of error-handling/refcounting bugs
* [cdefer.benchmark_results](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/44bbc6af7332d51840c9c1e1dcac6167b92a9377.benchmark_results) (2554 bytes) - added by Peaker on 2007-06-29 00:16:58Z - The benchmark results on my machine
* [cdefer.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/8ac9640d2de4e79f4f1ba49f44a84ac4671130dc.patch) (35184 bytes) - added by Peaker on 2007-06-29 00:18:27Z - Fixed a reentrancy bug (runCallbacks->callback->addCallbacks->runCallbacks). Added relevant test cases. Fixed use of defer.setTimeout in twisted.mail
* [cdefer.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/2e35ef293f6252bea7c214e0bc5ad8e2857ca84a.patch) (35311 bytes) - added by Peaker on 2007-06-29 00:27:41Z - Fixed a reentrancy bug (runCallbacks->callback->addCallbacks->runCallbacks). Added relevant test cases. Fixed use of defer.setTimeout in twisted.mail
* [cdefer-2245.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/854391fd10faa8cb6e838735e880c91bf959a6ff.patch) (504 bytes) - added by Peaker on 2007-07-03 11:44:19Z - Use int instead of Py_ssize_t as a PyList index.
* [cdefer.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/4360df8cb090f8e79b8b0eae6d568c3be16471a2.patch) (1908 bytes) - added by Peaker on 2007-07-04 10:50:51Z - Fixes setDebugging/setTimeout tests
* [cdefer.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/2cb2140105b06fa994e6579557b21da7efb1be78.patch) (1892 bytes) - added by Peaker on 2007-07-04 11:03:05Z - Use test.skip = '...' instead of in-test logic.
* [cdefer.6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/43915c40a4b8f7a9ceef8c65069bb8de6d390241.patch) (9938 bytes) - added by Peaker on 2007-07-04 22:23:46Z - Add real debugging support, the previous approach was completely broken
* [cdefer.7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/6fec1cf786ea78e586e08b2d5b3ac6e6ffef9a49.patch) (9940 bytes) - added by Peaker on 2007-07-04 23:01:05Z - Fix a bug in the way fail objects were passed to Failure in errback()
* [properties.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/ddd9c48220a76048597a0e575ae3ff10e48b7b3b.py) (1705 bytes) - added by Peaker on 2007-09-21 19:58:39Z - 
* [test_properties.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/2c2f1cc652ad9d2c3073fda78956bb37c20fe216.py) (2057 bytes) - added by Peaker on 2007-09-21 19:59:09Z - 
* [def.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/25e59387c6a3b3a97c1418cc4a62b77b4f29676b.py) (382 bytes) - added by redbaron on 2008-09-12 00:55:01Z - Minimal test-case shows huge memory leaks
* [cdef-memleak.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf7/cf72565d2a62067e4e33e16d9e81e366ad08dd54/5ede90e3393df03c74bc15c8c0cd1b3c1f4babe4.py) (357 bytes) - added by redbaron on 2008-09-12 09:11:35Z - Even simplier test which shows memory leaks

<details><summary>Searchable metadata</summary>

```
trac-id__2245 2245
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_cdefer_2245_4 branches/cdefer-2245-4
branch_author__ 
status__new new
resolution__ 
component__core core
keywords__ 
time__1164127046000000 1164127046000000
changetime__1450482154718768 1450482154718768
version__None None
owner__ 
cc__therve cc__spiv cc__peaker cc__exarkun cc__ed@... cc__thijs cc__vultaire cc__mkeller@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set status to `closed`|
|-|-|

This kind of investigation may be useful, but we need a performance benchmark suite for Deferreds before it can be _productive_, I think.

James Knight has previously implemented Deferred in Pyrex.  The reason this isn't included in Twisted is that we just have no reliable, accurate way to measure what the performance changes of this are.  The same applies to the attached cdefer.c.

In the interest of reducing the ever rising tide of tickets, I am closing this one, since it seems very unlikely to me that the work necessary to resolve it in a useful way is actually going to get done.  Creating this as a low priority ticket was the right kind of idea, but what trac really needs is a whole separate category of things that might be interesting but are unlikely to happen.

If you (therve) or anyone else would like to contribute a benchmark which makes it possible to determine the utility of a Pyrex or C implementation of Deferred, _please_ feel free to re-open this ticket: this is _not_ a rejection of the idea of a more optimized Deferred implementation.

Deriving something from the code which showed a 5% - 20% speedup might be a good way to start on such a benchmark.  It's important to cover all uses of Deferreds here.  It's also okay if some parts show a slowdown, as long as the overall improvement is enough.  Of course, it's better if everything is faster. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

Thanks a lot for your comments. A benchmark tool would be awesome, I'm putting this idea on my hope-i'll-have-some-time-one-day todo list. The main problem is to test **real** code, not some quick benchmark around Deferred.addCallback. It could even be a great buildbot metric of Twisted (for problem like [#11063](https://github.com/twisted/twisted/issues/11063)).

Until then I'll try to close some tickets :).

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_6"></a>|@Peaker commented|
|-|-|

* I added test cases for a bunch of bugs that existed.
* I added benchmarks for deferred behaviours:
  * With various N values (Which indeed discover that the old deferred has much worse-than-linear behaviour in various places, the C one is linear)
  * Benchmarks have good coverage of the .py implementation (except for chainDeferred, which is uninteresting to benchmark, debug-stuff which do not exist in the C version, and setTimeout code which also is not implemented in cDeferred)
* Reviewed the C code and fixed a lot of error handling/refcounting bugs (All of them, hopefully).
* Added the "Unhandled Error in Deferred" mechanism, so those are still reported in the same way (and the relevant tests pass, too) with cdefer.

Notes:
* The C deferred finally dropped support for the deprecated setTimeout mechanism.
* With the C deferred, setDebugging/getDebugging "works", but does not do anything. This fails a few debug-info tests (testAlreadyCalledDebug_[CE][CE]) that verify some debug output.
* Unhandled Error in Deferred still works.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @Peaker|
|-|-|

I applied the patch to the branch cdefer-2245.

I haven't really looked at the code, but I did notice that the attached cdefer.c doesn't build against against Python 2.3 or Python 2.4.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @Peaker|
|-|-|

I applied the patch on the branch and had a quick look. The modifications you made looks reasonable.

Unfortunately to be reviewed all the tests must pass, and for now the tests on debug and setTimeout doesn't pass. The latter could be skipped I guess since setTimeout is deprecated, but we have to find a solution to the former.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_11"></a>|@Peaker removed owner|
|-|-|

Replying to [therve](#note_10):
> I applied the patch on the branch and had a quick look. The modifications you made looks reasonable.
>
> Unfortunately to be reviewed all the tests must pass, and for now the tests on debug and setTimeout doesn't pass. The latter could be skipped I guess since setTimeout is deprecated, but we have to find a solution to the former.

I added a new patch that passes all of defer's tests:

I set setTimeout tests to only run if hasattr(Defer, "setTimeout") (I am not sure if this is better than actually removing the tests).

I changed setDebugging to only apply to newly created Deferreds, by using the Py-deferred when debugging is set to True, rather than the C deferred one.  This allows for backwards compatibility with debug, the tests to pass, at the expense of performance in those cases.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

I applied latest patch on the branch, and made some cleanups. That begins to look cool :).

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_14"></a>|@jyknight commented|
|-|-|

/me wonders upon the purpose of writing it directly in C rather than pyrex.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

The main purpose of Pyrex is to wrap an existing library. For an optimization, I don't think that's the best choice. Furthermore, memory/GC management with Pyrex can be problematic.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_16"></a>|@jyknight commented|
|-|-|

Replying to [therve](#note_15):
> The main purpose of Pyrex is to wrap an existing library. For an optimization, I don't think that's the best choice.

If the C implementation is sufficiently faster, that could be a good reason. But it seems to me, speed being equal, the more python-like pyrex should be preferred where possible. Or rpython

> Furthermore, memory/GC management with Pyrex can be problematic.

Compared to *C*?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

Replying to [jknight](#note_16):

> If the C implementation is sufficiently faster, that could be a good reason. But it seems to me, speed being equal, the more python-like pyrex should be preferred where possible. Or rpython

Well, yeah, of course, but I really think that the C version is faster than the Pyrex one, which tends to add a lot of useless generated code. We could try the benchmark on both version.

> > Furthermore, memory/GC management with Pyrex can be problematic.
>
> Compared to *C*?

Pyrex has (had ?) some problems without object deallocation for example. At least with C everything is explicit.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|spiv commented|
|-|-|

Replying to [therve](#note_17):
>
> Well, yeah, of course, but I really think that the C version is faster than the Pyrex one, which tends to add a lot of useless generated code. We could try the benchmark on both version.

Without actually measuring, statements about how fast they are are worthless.  Guessing the actual measurable difference in this sort of situation is notoriously inaccurate.

So benchmark first, then you can argue about whether the various tradeoffs are worth it.

FWIW, unless the difference in performance is stellar, I'd lean towards whichever has the more readable code, which presumably means pyrex.  But I want to see a benchmark...

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_19"></a>|@Peaker commented|
|-|-|

I initially worked on the Pyrex one from foom.

I spent many many hours trying to debug a GC-related crash that occured with pyrex deferred objects and circular references, when therve asked me if I had tried his c deferred implementation.

I am not sure at all which one is faster, but here is the comparison as I see it:

Pyrex pro's:
 1. Pyrex code is more readable and much easier to maintain
 1. Pyrex code provides more useful traceback information in its exceptions (builds traceback objects)

C pro's:
 1. C code is easier to build
 1. C code has more "transparent" memory management/refcounting semantics. With Pyrex, I am always puzzled about how refcounts are managed (e.g: If I cast a <void*> to an <object> do I get a new reference? A burrowed one?)

While I do believe that ease of building is less important, and that the first Pyrex point (easier maintenance) is highly important (it took hours to add the equivalent of a few Python/Pyrex lines to the C one), I just couldn't get the thing to work correctly, memory-management wise.

C memory management is not easy at all, but I believe my patch does all of the refcounting things correctly.  I could get it to work with all the circular cases.

I think the comparison of Pyrex/C performance here is probably not an issue. They are both far far faster than the Python defer, and probably similarly so.


If someone can comment on the viability of RPython in terms of ease of build and the other aspects I mentioned, I would be grateful. I just don't know enough about the use of RPython so I haven't tried using it.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_20"></a>|@jyknight commented|
|-|-|

Well, I'm sure you noticed the comment in the code where I call PyGC_UnTrack(self) explicitly in __dealloc__ because pyrex doesn't do that automatically. So, yeah, pyrex isn't perfect: you still do need to know the underlying C memory semantics where pyrex's abstraction falls down.

But, I was able to run the twisted unit test suite with my pyrex version w/o crashes. Were you able to do that, but still ran into problems in your own code? Or did that crash for you too?

For ease of building: the generated C code should be distributed along with the .pyx.

For rpython: I don't really know anything about it either, just threw it out there to confuse things further.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_21"></a>|@Peaker commented|
|-|-|

Replying to [jknight](#note_20):
> Well, I'm sure you noticed the comment in the code where I call PyGC_UnTrack(self) explicitly in __dealloc__ because pyrex doesn't do that automatically. So, yeah, pyrex isn't perfect: you still do need to know the underlying C memory semantics where pyrex's abstraction falls down.
>
> But, I was able to run the twisted unit test suite with my pyrex version w/o crashes. Were you able to do that, but still ran into problems in your own code? Or did that crash for you too?
>
> For ease of building: the generated C code should be distributed along with the .pyx.
>
> For rpython: I don't really know anything about it either, just threw it out there to confuse things further.
>

Yes, I have noticed that GC_Untrack call and tried using it.

The bug I had was with an adjusted test of defer's tests. That test may not have existed back when the Pyrex one was created, because it tried assigning into a deferred object's "_d" attribute which did not exist.
I converted it to simply assign into "result" which is allowed, and under undeterministic circumstances (When the test is run after some others, but not alone) it caused the GC_Untrack call to crash as the GC header in the object, iirc, was corrupt.

I am not remembering the details exactly, but I do know that I spent less time getting the C one to work, than I did trying to get the Pyrex one to work.

Defer seems relatively stable, so I think that what works correctly *now*, and has all the features, should matter more.

Also, therve wrote cdefer.c with good style, which makes the code reasonably maintainable, even if it is C.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_22"></a>|@jyknight commented|
|-|-|

The test in question indeed did not exist: test_defer.py:LogTestCase.test_errorLogWithInnerFrameCycle.

It'd be nice if someone could come up with another way to write that test that doesn't require writing random attributes or writing to .result. 

(I had considered making the pyrex deferred not expose any of the attributes but result, and making result be readonly. Might still be a good idea..)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

Replying to [jknight](#note_22):

> (I had considered making the pyrex deferred not expose any of the attributes but result, and making result be readonly. Might still be a good idea..)

I strongly disagree with the idea of making the C Deferred _in any way_ semantically distinct from the Python Deferred.  Even the "private" attributes, which nobody should ever access, are very occasionally helpful for debugging, and the last thing that I want is for the C Deferred to be even harder to debug.

However, I do think agree Deferred has way too many public attributes, which I doubt very much code uses.  I would be happy to see these start going away [in a gradual and compatibility-friendly way](http://twistedmatrix.com/trac/wiki/CompatibilityPolicy), but I think that should remain separate from this issue.

(I am not totally inflexible on the idea that the new C deferred will not support stuff that is deprecated.  However, I am adamant that the deprecations must be done in Python **first** before we even start talking about dropping features in the C version.)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_25"></a>|@jml set owner to @therve|
|-|-|


 * It is unclear from this ticket whether the cdefer-2245 branch is up-to-date and authoritative.
 * Without reading the discussion in-depth, it is difficult to determine whether all of the design and policy questions have been resolved.
 * Several patches and comments have been uploaded since the ticket was marked 'review'.

As such, I believe this ticket has _not_ reached the code review stage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @Peaker|
|-|-|

Peaker is the author of this code.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_27"></a>|@Peaker removed owner|
|-|-|

You are right, it was indeed not clear.

The cdefer-2245 branch is the up-to-date branch of this ticket.

The modifications that were made after it was set for review were fixes for problems that were discovered after setting for review, but it has been stable since.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_28"></a>|@jml commented|
|-|-|

Are there benchmarks? Have glyph's objections been resolved?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_29"></a>|@Peaker commented|
|-|-|

Replying to [jml](#note_28):
> Are there benchmarks? Have glyph's objections been resolved?
Yes. I am not sure what glyph's objection was.

If it was about allowing access to the deferred's "callbacks" attribute, then its not available in the C version. Should it be made available? It should probably be private and thus unavailable. How do you deprecate access to an attribute?  I am not sure it makes a lot of sense to do this in the C deferred.

I think if people still depend on silly things like accessing "callbacks" the right way to address that is some kind of way to tell Twisted to use the Python deferred in broken code.

Currently that way is editing defer.py :-) But maybe some better way should be used?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_30"></a>|@Peaker commented|
|-|-|

Report of differences between defer with/without cdefer:
 * The callbacks attribute refers to a list of a different structure. In the Py one its [ ((callback, cbargs, cbkw), (errback, ebargs, ebkw))* ]. In the C one its [ (callback, cbargs, cbkw, errback, ebargs, wbkw)* ].
 * There is no 'debug' class attribute in cdefer.Deferred. Note that the proper way of getting/setting debug is via getDebugging/setDebugging which are implemented in cdefer.
 * Deferred is an old-style class. cdefer.Deferred is a new-style class. This also affects DeferredList or anyone else who inherits from it.
 * timeout was finally removed in cdefer, with it the 'timeoutCall' attribute was removed.
 * The private _debugInfo attribute is not accessible
 * 'callbacks', 'paused', 'called' are read-only attributes in cdeferred but writable in Deferred.

I believe this sums up all of the differences.
Which of these are acceptable, and which should be corrected?

Replying to [Peaker](#note_29):
> Replying to [jml](#note_28):
> > Are there benchmarks? Have glyph's objections been resolved?
> Yes. I am not sure what glyph's objection was.
>
> If it was about allowing access to the deferred's "callbacks" attribute, then its not available in the C version. Should it be made available? It should probably be private and thus unavailable. How do you deprecate access to an attribute?  I am not sure it makes a lot of sense to do this in the C deferred.
>
> I think if people still depend on silly things like accessing "callbacks" the right way to address that is some kind of way to tell Twisted to use the Python deferred in broken code.
>
> Currently that way is editing defer.py :-) But maybe some better way should be used?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_31"></a>|@Peaker commented|
|-|-|

Another difference I missed:
 * Class attributes are used for defaults in the .py one, and don't exist in the type object.

Replying to [Peaker](#note_30):
> Report of differences between defer with/without cdefer:
>  * The callbacks attribute refers to a list of a different structure. In the Py one its [ ((callback, cbargs, cbkw), (errback, ebargs, ebkw))* ]. In the C one its [ (callback, cbargs, cbkw, errback, ebargs, wbkw)* ].
>  * There is no 'debug' class attribute in cdefer.Deferred. Note that the proper way of getting/setting debug is via getDebugging/setDebugging which are implemented in cdefer.
>  * Deferred is an old-style class. cdefer.Deferred is a new-style class. This also affects DeferredList or anyone else who inherits from it.
>  * timeout was finally removed in cdefer, with it the 'timeoutCall' attribute was removed.
>  * The private _debugInfo attribute is not accessible
>  * 'callbacks', 'paused', 'called' are read-only attributes in cdeferred but writable in Deferred.
>
> I believe this sums up all of the differences.
> Which of these are acceptable, and which should be corrected?
>
> Replying to [Peaker](#note_29):
> > Replying to [jml](#note_28):
> > > Are there benchmarks? Have glyph's objections been resolved?
> > Yes. I am not sure what glyph's objection was.
> >
> > If it was about allowing access to the deferred's "callbacks" attribute, then its not available in the C version. Should it be made available? It should probably be private and thus unavailable. How do you deprecate access to an attribute?  I am not sure it makes a lot of sense to do this in the C deferred.
> >
> > I think if people still depend on silly things like accessing "callbacks" the right way to address that is some kind of way to tell Twisted to use the Python deferred in broken code.
> >
> > Currently that way is editing defer.py :-) But maybe some better way should be used?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

Okay here are some opinions from me:

  * The different callbacks structure is unfortunate.  If there is a good reason for this, maybe it's okay.  If it's just an accident, then it should probably be made consistent with Deferred.
  * It would be best if there were deprecated support for setting cdefer.Deferred.debug.  defer.Deferred.debug should also actually be deprecated too.  The actual deprecation part of this can be a separate task, though.  I don't know how you implement cdefer.Deferred.debug, though.
  * I guess it's technically possible to have a classic extension type?  I'm not really sure though.  I don't think I care about this change though.  Maybe we should make it more explicit that subclassing Deferred isn't useful.
  * Timeout stuff has been deprecated since 2.0.  We should also drop it from defer.py.
  * _debugInfo can be considered an implementation detail of defer.py, cdefer doesn't need to provide it.
  * I'm a little fuzzy on whether it's okay for those attributes to be read-only.  I feel like it is, but maybe that's just a subjective thing, and actually we should follow the deprecation policy for this change.

Sorry for not providing absolutely concrete responses.  Maybe if I think about it a while longer I'll be more certain what the correct choices are.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_33"></a>|@Peaker commented|
|-|-|

I deprecated the paused/callbacks attributes now.
I also fixed the callbacks list to share the same structure as the Py Deferred list.

Here's an updated list (Now ordered from most substantial to least, IMO):
 * 'callbacks', 'paused', 'called' are read-only attributes in cdeferred but writable in Deferred (Now callbacks/paused are deprecated too).
 * There is no 'debug' class attribute in cdefer.Deferred. Note that the proper way of getting/setting debug is via getDebugging/setDebugging which are implemented in cdefer.
 * Deferred is an old-style class. cdefer.Deferred is a new-style class. This also affects DeferredList or anyone else who inherits from it.
 * Class attributes are used for defaults in the .py one, and don't exist in the type object.

I think its reasonable to assume there's a concensus that these are ok:
 * timeout was finally removed in cdefer, with it the 'timeoutCall' attribute was removed.
 * The private _debugInfo attribute is not accessible

Replying to [Peaker](#note_31):
> Another difference I missed:
>  * Class attributes are used for defaults in the .py one, and don't exist in the type object.
>
> Replying to [Peaker](#note_30):
> > Report of differences between defer with/without cdefer:
> >  * The callbacks attribute refers to a list of a different structure. In the Py one its [ ((callback, cbargs, cbkw), (errback, ebargs, ebkw))* ]. In the C one its [ (callback, cbargs, cbkw, errback, ebargs, wbkw)* ].
> >  * There is no 'debug' class attribute in cdefer.Deferred. Note that the proper way of getting/setting debug is via getDebugging/setDebugging which are implemented in cdefer.
> >  * Deferred is an old-style class. cdefer.Deferred is a new-style class. This also affects DeferredList or anyone else who inherits from it.
> >  * timeout was finally removed in cdefer, with it the 'timeoutCall' attribute was removed.
> >  * The private _debugInfo attribute is not accessible
> >  * 'callbacks', 'paused', 'called' are read-only attributes in cdeferred but writable in Deferred.
> >
> > I believe this sums up all of the differences.
> > Which of these are acceptable, and which should be corrected?
> >
> > Replying to [Peaker](#note_29):
> > > Replying to [jml](#note_28):
> > > > Are there benchmarks? Have glyph's objections been resolved?
> > > Yes. I am not sure what glyph's objection was.
> > >
> > > If it was about allowing access to the deferred's "callbacks" attribute, then its not available in the C version. Should it be made available? It should probably be private and thus unavailable. How do you deprecate access to an attribute?  I am not sure it makes a lot of sense to do this in the C deferred.
> > >
> > > I think if people still depend on silly things like accessing "callbacks" the right way to address that is some kind of way to tell Twisted to use the Python deferred in broken code.
> > >
> > > Currently that way is editing defer.py :-) But maybe some better way should be used?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_34"></a>|@Peaker commented|
|-|-|

Replying to [exarkun](#note_32):
> Okay here are some opinions from me:
>
>   * The different callbacks structure is unfortunate.  If there is a good reason for this, maybe it's okay.  If it's just an accident, then it should probably be made consistent with Deferred.
I fixed it to be consistent. There's a minor difference (None instead of passthru) but user code really shouldn't be affected. (1. They should not have looked at defer.callbacks in the first place, 2. They should not compare with passthru, really).


>   * It would be best if there were deprecated support for setting cdefer.Deferred.debug.  defer.Deferred.debug should also actually be deprecated too.  The actual deprecation part of this can be a separate task, though.  I don't know how you implement cdefer.Deferred.debug, though.
This is difficult, because it requires using a metaclass. I'm attempting to see how horrible it really is though.


>   * I guess it's technically possible to have a classic extension type?  I'm not really sure though.  I don't think I care about this change though.  Maybe we should make it more explicit that subclassing Deferred isn't useful.
I don't think that the differences between old-style and new-style classes should matter to any reasonable code. Subclassing still works, by the way, it will just behave a bit differently (and subclasses are those that are more likely to notice any differences).

>   * Timeout stuff has been deprecated since 2.0.  We should also drop it from defer.py.
Okay. That's a matter for another ticket, though.


>   * _debugInfo can be considered an implementation detail of defer.py, cdefer doesn't need to provide it.
That's settled then :-)

>   * I'm a little fuzzy on whether it's okay for those attributes to be read-only.  I feel like it is, but maybe that's just a subjective thing, and actually we should follow the deprecation policy for this change.
Its not hard to allow them to be writable. I thought it simply doesn't "make sense". I guess the correct thing to do is to allow it with a deprecation warning, for now.


> Sorry for not providing absolutely concrete responses.  Maybe if I think about it a while longer I'll be more certain what the correct choices are.

Thanks for the reply!

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_35"></a>|@Peaker commented|
|-|-|

 * Added the "debug" attribute with deprecation as alternative to getDebugging/setDebugging.
 * Added write access to deprecated "paused" attribute with deprecation warnings.

The differences that still remain:
 * called and callbacks are read-only. I can allow modifying them but that's not tested in the current tests and I don't think that much code would be broken by this.
 * Deferred is an old-style class. cdefer.Deferred is a new-style class. This also affects DeferredList? or anyone else who inherits from it. 
 * Class attributes are used for defaults in the .py one, and don't exist in the type object.

In my opinion, the 2nd and 3rd differences are non-issues. The first one is a matter of arbitrary choice - which I made in this way. If you disagree, please explain why.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_36"></a>|@therve set owner to @Peaker|
|-|-|

There is warning in the compilation which is suspect to me:
```
twisted/internet/cdefer/cdefer.c: In function 'cdefer_Deferred_get_callbacks':
twisted/internet/cdefer/cdefer.c:848: warning: implicit declaration of function 'PyErr_WarnEx'
twisted/internet/cdefer/cdefer.c: At top level:
twisted/internet/cdefer/cdefer.c:953: warning: initialization from incompatible pointer type
```

It seems it has been added in 2.5, we need to support 2.3 and 2.4. The consequence is a failing test.

The changes to twisted.mail.alias must be reverted.

The test methods should be named 'test_foo', not 'testFoo'. They also need docstrings.

If the access to attributes is done in cdefer, it must be done in defer too. That may be done in another ticket, I'm not sure yet.

I'm not sure having specific cases for cdefer is sane. The goal is to have the same behavior (this is related to previous remark).

The benchmark tool have an unused import functools. It would probably need some docstrings too.

That's it for now!

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_37"></a>|@Peaker commented|
|-|-|

I toyed with the idea of using properties to deprecate things in the Python Deferred too, before realizing its probably best to just have the C one behave like the old Python one and allow writing on the exposed private attributes of the past. So I am attaching a utility module for twisted.python to allow creation of properties that deprecate attribute access.
Due to limitations in "property" in Python, it is probably more useful on new-style classes, if its ever used.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_38"></a>|@Peaker commented|
|-|-|

Thanks for the review. I addressed your comments:

Replying to [therve](#note_36):
> There is warning in the compilation which is suspect to me:
> ```
> twisted/internet/cdefer/cdefer.c: In function 'cdefer_Deferred_get_callbacks':
> twisted/internet/cdefer/cdefer.c:848: warning: implicit declaration of function 'PyErr_WarnEx'
> twisted/internet/cdefer/cdefer.c: At top level:
> twisted/internet/cdefer/cdefer.c:953: warning: initialization from incompatible pointer type
> ```
>
> It seems it has been added in 2.5, we need to support 2.3 and 2.4. The consequence is a failing test.

I removed the deprecations and the call to PyErr_WarnEx with them. There was also another silly warning I fixed. It compiles without warnings now.

> The changes to twisted.mail.alias must be reverted.
Done. Though now they fail because cdefer has no setTimeout. More on that below.

> The test methods should be named 'test_foo', not 'testFoo'. They also need docstrings.
Fixed all of test_defer names to be test_foo. Also added docstrings to all of the tests I added, and to some tests that were already there and helper methods. Adding docstrings to _all_ of the defer tests is probably out of the scope of this ticket though.

> If the access to attributes is done in cdefer, it must be done in defer too. That may be done in another ticket, I'm not sure yet.
Ok, I changed it to allow access to attributes in c defer in the same way, with writability and no deprecations.  Another ticket can be about the privatizing of too many publicly-named attributes of deferred.

> I'm not sure having specific cases for cdefer is sane. The goal is to have the same behavior (this is related to previous remark).
Removed those, now that attributes behave the same.

> The benchmark tool have an unused import functools. It would probably need some docstrings too.
Fixed.

> That's it for now!

One problem remains, as mentioned above, that of setTimeout.

setTimeout must finally be removed for tests to pass with this ticket, because it is not and will not be implemented in the C deferred.

If its done in another ticket, it forms a dependency on it, and this dependency should be mentioned here.

What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_40"></a>|@therve set owner to @Peaker|
|-|-|

 * test_reentrantReadCallbacks is failing with ```PyDeferred```
 * I'm not sure the change in defer.py about ```DebugInfo``` instantiation is relevant here. Can you explain the change?
 * Please don't make changes like replacing testMethod by test_method. This is what we want for new methods, but big changes like that just breaks every other branch around modifying test_defer.
 * the docstrings must be written
```
"""
docstring
"""
```
 even if it's on one line
 * there are trailing whitespaces in test_defer

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_41"></a>|@Peaker commented|
|-|-|

Thanks for the review! Comments in-place:

Replying to [therve](#note_40):
>  * test_reentrantReadCallbacks is failing with ```PyDeferred```
Yeah, that's a bug in defer.py (You were not using the c defer one). I fixed it in [#2849](https://github.com/twisted/twisted/issues/2849) and waiting for that to merge before submitting this for review again.

>  * I'm not sure the change in defer.py about ```DebugInfo``` instantiation is relevant here. Can you explain the change?
The change is made to make it easier for the C deferred to use the DebugInfo object. Adding the string prefix from Python is trivial and from C a bit more difficult...

>  * Please don't make changes like replacing testMethod by test_method. This is what we want for new methods, but big changes like that just breaks every other branch around modifying test_defer.

Ok, I reverted the name changes to existing test methods.

>  * the docstrings must be written
> ```
> """
> docstring
> """
> ```
>  even if it's on one line

Fixed.

>  * there are trailing whitespaces in test_defer
Fixed.

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_42"></a>|peaker commented|
|-|-|

(In [21544]) Initial merge-forward, not ready yet. refs [#2245](https://github.com/twisted/twisted/issues/2245)

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_43"></a>|peaker commented|
|-|-|

(In [21547]) Passes the new defer tests, with the new [#2849](https://github.com/twisted/twisted/issues/2849) behavior. refs [#2245](https://github.com/twisted/twisted/issues/2245)

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_44"></a>|peaker commented|
|-|-|

(In [21554]) Refactor test_reentrantReadCallbacks. refs [#2245](https://github.com/twisted/twisted/issues/2245)

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_45"></a>|peaker commented|
|-|-|

(In [21555]) Remove redundant test (was added in [#2849](https://github.com/twisted/twisted/issues/2849)). Fix comment. refs [#2245](https://github.com/twisted/twisted/issues/2245)

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_46"></a>|@Peaker commented|
|-|-|

I think its basically ready for review, but cannot be merged because twisted.mail.alias still uses setTimeout (arrg!). [#2729](https://github.com/twisted/twisted/issues/2729) is about fixing it, but I am not sure why all that fuss is necessary, as the attached http://twistedmatrix.com/trac/attachment/ticket/2245/cdefer.3.patch fixes it as well (and that specific fix can be put into this branch).

Please give it a look anyhow.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_49"></a>|@exarkun set owner to @Peaker<br>@exarkun set status to `new`|
|-|-|

  * doc/core/benchmarks/defer.py
    * missing:
      * copyright notice
      * module docstring (okay it's not really a module, but there should be some overall description of intent)
    * docstrings should be ```"""\ntext\n"""```, even if they fit on one line (a docstring is a very important thing, it merits the visibility this style provides it)
    * the naming convention is ```camelCase```, not ```under_scores```
    * Isn't ```pause_unpause``` basically the same as ```instantiate_add_callbacks_before_result```?  ```Deferred.pause``` and ```Deferred.unpause``` don't get used very frequently.
    * It might be interesting to see ```Deferred.callback``` benchmarked on different sized callbacks lists separately from ```Deferred.addCallbacks```.  ```instantiate_add_callbacks_no_result``` already does the reverse.
  * doc/core/benchmarks/timer.py
    * changes seem more or less irrelevant
    * ```camelCase```, not ```under_scores```.
  * twisted/test/test_defer.py
    * Thanks for the additional docstrings
    * ```_verify_cb``` should probably be ```_verifyCallbackEntry``` or something.
    * in docstrings, use I{} or B{} for emphasis (epytext markup), C{} for code literals, L{} for strings which identify a python object (eg, L{Deferred.callbacks}).
    * some trailing whitespace in ```test_reentrantReadCallbacks```
    * ```test_addCallbackErrbackWithoutArgs``` could be using ```assertRaises```
  * twisted/internet/cdefer.c
    * cdefer_Deferred!__runCallbacks has this line: ```newArgs = Py_BuildValue("(O)", self->result);```.  A few lines below, if ```PySequence_InPlaceConcat``` fails, the function returns without cleaning up ```newArgs```.  That's a leak, right?
  * bugs I found
    * 
```python
>>> from twisted.internet.defer import Deferred
>>> d = Deferred()
>>> d.callbacks = [None]
>>> d.callback(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in ?
SystemError: error return without exception set
>>>
```
    * 
```python
>>> from twisted.internet.defer import Deferred
>>> d = Deferred()
>>> d.callbacks = [0]
>>> d.callback(None)
Segmentation fault
```


I should really spend more time looking at the C, but that's a bunch of stuff to start with.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_50"></a>|@Peaker commented|
|-|-|

Replying to [exarkun](#note_49):
>   * doc/core/benchmarks/defer.py
>     * missing:
>       * copyright notice
Done

>       * module docstring (okay it's not really a module, but there should be some overall description of intent)
Done

>     * docstrings should be ```"""\ntext\n"""```, even if they fit on one line (a docstring is a very important thing, it merits the visibility this style provides it)
Done

>     * the naming convention is ```camelCase```, not ```under_scores```
Done

>     * Isn't ```pause_unpause``` basically the same as ```instantiate_add_callbacks_before_result```?  ```Deferred.pause``` and ```Deferred.unpause``` don't get used very frequently.
Don't know, the benchmark isn't trying to assume much about the implementation. It is another API so its tested (I remember trying to cover as many of the APIs as reasonable). Benchmark is showing slightly differing results, too.

>     * It might be interesting to see ```Deferred.callback``` benchmarked on different sized callbacks lists separately from ```Deferred.addCallbacks```. ```instantiate_add_callbacks_no_result``` already does the reverse.

I guess you could just subtract the time of instantiateAddCallbacksNoResult from the time of instantiateAddCallbacksBeforeResult...

>   * doc/core/benchmarks/timer.py
>     * changes seem more or less irrelevant
The reason I changed timer.py is to remove the getattr(time, 'time') call from the time measurement. It may be insignificant, but if the iteration count is small, it can have a small effect.

>     * ```camelCase```, not ```under_scores```.
Done.

>   * twisted/test/test_defer.py
>     * Thanks for the additional docstrings
No problemos, amigos.

>     * ```_verify_cb``` should probably be ```_verifyCallbackEntry``` or something.
Done.

>     * in docstrings, use I{} or B{} for emphasis (epytext markup), C{} for code literals, L{} for strings which identify a python object (eg, L{Deferred.callbacks}).

I tried fixing that but I am not sure I did it correctly, as I am not sure when to use C{} and when to use L{}. Why is "Deferred" not wrapped in C{} or L{} in all of the docstrings, for example?

>     * some trailing whitespace in ```test_reentrantReadCallbacks```
Done.

>     * ```test_addCallbackErrbackWithoutArgs``` could be using ```assertRaises```
It is using assertRaises but comes with a comment:
```
# assertRaises gives a **{} to the function, whereas no arg at all
# gives a NULL to cdefer, so lets call it directly here:
```
that explains why it is being duplicated in there.

>   * twisted/internet/cdefer.c
>     * cdefer_Deferred!__runCallbacks has this line: ```newArgs = Py_BuildValue("(O)", self->result);```.  A few lines below, if ```PySequence_InPlaceConcat``` fails, the function returns without cleaning up ```newArgs```.  That's a leak, right?
Yeah :-(  That code could probably be broken into smaller more verifiable functions. But for now I'll fix that specifically.

>   * bugs I found
>   ....

Fixed. These bugs result from allowing the user to set 'callbacks', but perhaps it will be better to make 'callbacks', 'paused' and other attributes read only in the context of this ticket? I doubt anyone is depending on the ability to set these attributes on Deferred objects.

> I should really spend more time looking at the C, but that's a bunch of stuff to start with.

That's the heavy part of this ticket review...

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_51"></a>|@Peaker commented|
|-|-|

The current branch, for anyone wondering, is cdefer-2245-2

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_52"></a>|@Peaker commented|
|-|-|

Okay, I just imitate the behavior of pydeferred w.r.t handling of invalid callbacks lists. There's still some difference (some callbacks format failures will just jump to the errback chain in pydeferred but will raise through callback() or errback() in cdefer) but I believe that assigning callbacks is a corner case in itself, and the exact exception behaviour of such assignment that is even _wrong_ is not very important.

Mostly, the behavior is the same, anyhow, and there are no more segfaults. I also added test cases that include both of those previously crashing snippets.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_55"></a>|@therve commented|
|-|-|

Just for info, the branch needs to be merged forward.

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_56"></a>|peaker commented|
|-|-|

(In [21972]) Branching to 'cdefer-2245-3'

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_57"></a>|peaker commented|
|-|-|

(In [21974]) Branching to 'cdefer-2245-3'

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_58"></a>|@Peaker commented|
|-|-|

Replying to [peaker](#note_57):
> (In [21974]) Branching to 'cdefer-2245-3'

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_59"></a>|edsuom commented|
|-|-|

I tested cdefer-2245-3 with my [AsynCluster](http://foss.eepatents.com/AsynCluster) installation running on six CPU core nodes. The computing job involves a thousand or so interations with 1000 tasks per iteration, distributed over the six nodes on an as-available basis with [AsynQueue](http://foss.eepatents.com/AsynQueue) dispatching the tasks from a queue. Four of the nodes, the AsynCluster master process, and the python for the computing job itself ran on an Intel Core QX9650 overclocked to 3.6 GHz. The other two nodes ran on an AMD64 4200+. Most of the time for each of the 1000 tasks per iteration was spent in highly optimized Numpy calls and inline C for array processing.

The first trial was with cdefer-2245-3 as-is. For each module in the dependency code and the computing job with any significant deferred usage, I did a separate import of cdefer in a try:except loop, and then did ```defer.Deferred = cdefer.Deferred```.

For the second trial, I disabled the cdefer capability by renaming the ```cdefer.so``` file and ```cdefer``` directory in ```/usr/lib/python2.4/site-packages/twisted/internet```.

Here are the iteration times for each, with about a thousand iterations for each trial:

## C Deferreds
```
Min.    1st Quartile  Median  Mean    3rd Quartile    Max. 
0.150   0.630         0.840   0.767   0.950           1.400
```

## Python Deferreds
```
Min.    1st Quartile  Median  Mean    3rd Quartile    Max. 
0.1900  0.7300        0.8800  0.8033  0.9300          1.4000 
```

It is probably a _tiny_ bit better with C, but the improvement is marginal indeed for this real-world application. Too bad; I like the idea of optimizing this critical part of Twisted.

A possible next step will be to take dash's suggestion to work on this some more with a profiler.

**Comment by twisted-trac** [comment]
|<img alt="ncryptor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_60"></a>|ncryptor commented|
|-|-|

**posted on behalf of peaker**

Replying to [edsuom](#note_59):
> I tested cdefer-2245-3 with my [AsynCluster](http://foss.eepatents.com/AsynCluster) installation running on six CPU core nodes. The computing job involves a thousand or so interations with 1000 tasks per iteration, distributed over the six nodes on an as-available basis with [AsynQueue](http://foss.eepatents.com/AsynQueue) dispatching the tasks from a queue. Four of the nodes, the AsynCluster master process, and the python for the computing job itself ran on an Intel Core QX9650 overclocked to 3.6 GHz. The other two nodes ran on an AMD64 4200+. Most of the time for each of the 1000 tasks per iteration was spent in highly optimized Numpy calls and inline C for array processing.
>
> The first trial was with cdefer-2245-3 as-is. For each module in the dependency code and the computing job with any significant deferred usage, I did a separate import of cdefer in a try:except loop, and then did ```defer.Deferred = cdefer.Deferred```.
>
> For the second trial, I disabled the cdefer capability by renaming the ```cdefer.so``` file and ```cdefer``` directory in ```/usr/lib/python2.4/site-packages/twisted/internet```.
>
> Here are the iteration times for each, with about a thousand iterations for each trial:
>
> == C Deferreds ==
> ```
> Min.    1st Quartile  Median  Mean    3rd Quartile    Max.
> 0.150   0.630         0.840   0.767   0.950           1.400
> ```
>
> == Python Deferreds ==
> ```
> Min.    1st Quartile  Median  Mean    3rd Quartile    Max.
> 0.1900  0.7300        0.8800  0.8033  0.9300          1.4000
> ```
>
> It is probably a _tiny_ bit better with C, but the improvement is marginal indeed for this real-world application. Too bad; I like the idea of optimizing this critical part of Twisted.
>
> A possible next step will be to take dash's suggestion to work on this some more with a profiler.

I think this is not the main target for cdefer's optimization.

An app that already spends a lot of CPU time doing stuff and a little of its time doing Deferred stuff, will of course gain less than an app that spends almost all of its CPU time dispatching asynchronous calls and calling deferreds.

About your performance measurements: Are these real-time measurements, or CPU time measurements? In this context it probably makes more sense to measure CPU time.

Also, I think the min time probably represents the actual CPU performance best, as non-min times probably contain app-specific and other blocks/delays [including context switches or the specific disk's seek times, etc].

I believe the application-wide min time speed increase of 26% is definitely significant. If Deferreds runtime is no more than X% of your runtime, no matter how fast it will get, you cannot speed up your app by more than a factor of 100/(100-X).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_62"></a>|@therve commented|
|-|-|

(In [23213]) Branching to 'cdefer-2245-4'

**Comment by twisted-trac** [comment]
|<img alt="redbaron's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_63"></a>|redbaron commented|
|-|-|

I've attached minimal example of huge memory leaks using latest cdeferred implementation.

**Comment by twisted-trac** [comment]
|<img alt="redbaron's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_64"></a>|redbaron commented|
|-|-|

I've no C expirience at all but after reading docs and some experiments commenting out Py_INCREF(self->result); on line 728 seems to solve at least previously attached testcase.

**Comment by twisted-trac** [comment]
|<img alt="redbaron's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_65"></a>|redbaron commented|
|-|-|

I've discovered that memory leaks occurs on every cb or eb. Argument should be relatively big to make memleak noticible, I use Failure() instance for that purpose, I wrap it into my class to show that ordinary callbacks are affected to.

And yes, commenting Py_INCREF(self->result); on line 728 in cdefer.c definetely solves the problem.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_67"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_1):
> This kind of investigation may be useful, but we need a performance benchmark suite for Deferreds before it can be _productive_, I think.
>
> James Knight has previously implemented Deferred in Pyrex.  The reason this isn't included in Twisted is that we just have no reliable, accurate way to measure what the performance changes of this are.  The same applies to the attached cdefer.c.
>

I moved the defer.py from the branch for this ticket to the attachment of [#3704](https://github.com/twisted/twisted/issues/3704) for creating a deferred benchmark. Perhaps we should first complete that [ticket:3704 benchmark ticket] and then move on with this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_68"></a>|@thijstriemstra commented|
|-|-|

Replying to [therve](#note_2):
> It could even be a great buildbot metric of Twisted (for problem like [#11063](https://github.com/twisted/twisted/issues/11063)).

Running the benchmarks with buildbot is described in [#2201](https://github.com/twisted/twisted/issues/2201)

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_70"></a>|spiv commented|
|-|-|

See also [#3245](https://github.com/twisted/twisted/issues/3245), which has a benchmark (for memory consumption).

**Comment by twisted-trac** [comment]
|<img alt="vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_71"></a>|vultaire commented|
|-|-|

I'm quite curious about this branch; just pulled a copy locally.  I think for the app we're working on here, this may have a significant impact.

The server I've been working on does massive async calls, and on our last benchmark it appeared to spend over 25% of CPU time in the defer module.  About 11% was in inlineCallbacks and friends, and 12% was in the Deferred class itself.  (Counting method: sum of immediate CPU time for all functiosn in t.i.defer.py, using cProfile.  Note: we use MySQL with adbapi, so I think the threadpool threads could be skewing the results slightly.)

Seems there's supposedly a memory leak mentioned by redbaron, and other than that, this is still just pending review or something?  What's the status?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_72"></a>|@glyph commented|
|-|-|

Replying to [vultaire](#note_71):

> Seems there's supposedly a memory leak mentioned by redbaron, and other than that, this is still just pending review or something?  What's the status?

I could be wrong about this, but unless someone more informed feels like chiming in:

  * There's the memory leak you mentioned; that needs to be investigated and fixed.
  * I don't believe that the code in this branch implements `cancel()`, as recently added by [#10665](https://github.com/twisted/twisted/issues/10665).  That would need to be fixed.
  * I think we might be waiting on [#4593](https://github.com/twisted/twisted/issues/4593), especially so that the maximum-recursion-depth tests can be applied to this branch.

Merging can be a pain, so if you want to bring this up to date with trunk without svn commit access you probably want to just rebase the whole thing and make a new branch off of a recent trunk.  If you'd like to work on these issues, though, please feel free to attach patches or 'bzr send' output.

**Comment by twisted-trac** [comment]
|<img alt="vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_73"></a>|vultaire commented|
|-|-|

Not sure how much time I can commit, but I'd like to see this working.

I can comment on the memory leak issue above.  As redbaron mentioned, there was a PyINCREF(self->result) on line 728 in cdefer.c.  self->result was assigned from PyObject_Call, which looks like it does its own PyINCREF (http://docs.python.org/c-api/object.html#PyObject_Call).  Removing that PyINCREF looks like it fixes the problem.

The other issues, I can't strongly say I can commit the time right now, but I'll see what I can do.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_75"></a>|@exarkun commented|
|-|-|

Also, this is a big pile of C.  It's really unclear that it's going to be worth maintaining, even if you do all the work to fix the points Glyph mentioned.  If you can contribute a benchmark that reflects how your application uses Deferreds and demonstrates that this change is actually significantly helpful, that would go a long ways towards proving there's a reason we should undertake this maintenance burden.

**Comment by twisted-trac** [comment]
|<img alt="vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_76"></a>|vultaire commented|
|-|-|

That's fair.  I'll see what I can do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_77"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_75):
> Also, this is a big pile of C.  It's really unclear that it's going to be worth maintaining, even if you do all the work to fix the points Glyph mentioned.  If you can contribute a benchmark that reflects how your application uses Deferreds and demonstrates that this change is actually significantly helpful, that would go a long ways towards proving there's a reason we should undertake this maintenance burden.

I would have mentioned benchmarks first, but I thought that [#3704](https://github.com/twisted/twisted/issues/3704) already took care of that?  Not that I'd discourage vultaire from adding more benchmarks, of course :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_78"></a>|@exarkun commented|
|-|-|

Thanks for mentioning that ticket.  Any new benchmarks should be written in a way that play nicely with the ones that exist already, and clearly there's no point in overlapping with what's already covered.

Since [#3704](https://github.com/twisted/twisted/issues/3704) was resolved 15 months ago and this ticket is still open, the main conclusion I can draw is that the benchmarks it added don't show any compelling reason to bother with a C implementation of Deferred.  That may simply be because they benchmarks there aren't realistic (but it may also be because C Deferred doesn't speed up real programs).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_79"></a>|@therve removed owner|
|-|-|

The C extension _does_ speed up real programs, I've observed it in a few scenarios. Does it make a big difference? The most I've seen is about 20%, so I'm not such if it really worths the trouble. Maintaining C is hard. But the test coverage of !Deferred is not as bad is it used to be, especially with the tests in [#4593](https://github.com/twisted/twisted/issues/4593).

I guess this code needs a champion for the time being, but I'd rather try to tackle [#4593](https://github.com/twisted/twisted/issues/4593) if I find some time.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_80"></a>|@therve commented|
|-|-|

Gah, I'm not sure it's really worth the trouble.

**Comment by twisted-trac** [comment]
|<img alt="vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_81"></a>|vultaire commented|
|-|-|

Well, if it's not, we could end up just using it locally for our purposes.  That'd be fine; I can understand that in most cases this is likely not the bottleneck for a program.

That being said, I see a difference in a benchmark I wrote between the two, and I'll check out the benchmarks of [#3704](https://github.com/twisted/twisted/issues/3704) as well.  I'll agree that I would expect more from a C extension.

But I'm thinking there's a few places the extension could still be improved upon where more performance could be won.  If I can get something compelling together, I'll share it.

**Comment by twisted-trac** [comment]
|<img alt="vultaire's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_82"></a>|vultaire commented|
|-|-|

Bleh, rather vague last comment.  So far I've seen a 20+% difference between cdefer and defer in overall time in one benchmark I've done so far.  Not as high of a boost as I would expect for a C extension.

And of course, on a real system that's doing real work like database queries, caching, running some algorithms on the data, etc., the above 20% boost may translate to only 7% or less on overall time.  Probably not worth the effort as-is.  Hopefully it's possible to make this a bit faster...

**Comment by twisted-trac** [comment]
|<img alt="mkeller@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_83"></a>|mkeller@... commented|
|-|-|

I am working for a project that heavily uses Deferreds (viff.dk). Using an adapted version of the C extension, our benchmarks run 50 to 100 percent faster. Therefore, I would like to use the code, but there is no copyright notice in the patches. Is the license the same as for the whole project? Who do I have to credit?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by glyph** [comment]
Perhaps, functionally, a duplicate: #12124
